fusedrug/data/protein/structure/flexible_align_chains_structure.py

```diff
@@ -0,0 +1,293 @@
+    )
+
+    # dynamic_indices = dynamic_indices[:50]
+    # static_indices = static_indices[:50]
+
+    # extract seq-level matching atoms coordinates
+    dynamic_matching = _apply_indices(dynamic_concat, dynamic_indices)
+    static_matching = _apply_indices(static_concat, static_indices)
+
+    # calculate the rigid transformation to translate from the starting pose of the dynamic onto the static
+
+    combined_mask = np.logical_and(
+        dynamic_matching["atom14_gt_exists"].astype(bool),
+        static_matching["atom14_gt_exists"].astype(bool),
+    )
+    # orig_atom_pos_shape = dynamic_matching["atom14_gt_positions"].shape
+    use_for_static = static_matching["atom14_gt_positions"]
+    use_for_dynamic = dynamic_matching["atom14_gt_positions"]
+    if backbone_only_based:
+        use_for_static = use_for_static[:, :4, ...]
+        use_for_dynamic = use_for_dynamic[:, :4, ...]
+
+    _, rmsd, rot_matrix, trans_matrix = superimpose(
+        use_for_static.reshape(-1, 3),
+        use_for_dynamic.reshape(-1, 3),
+        combined_mask.reshape(-1),
+        verbose=True,
+    )
+
+    assert rot_matrix.shape == (1, 3, 3)
+    rot_matrix = rot_matrix[0]
+
+    assert trans_matrix.shape == (1, 3)
+    trans_matrix = trans_matrix[0]
+
+    assert len(rmsd.shape) == 0
+
+    if rmsd > 6.0:
+        warn(
+            f"flexible_align_chains_structure: got a pretty high rmsd={rmsd} in alignment. Either the structures are very different or the sequence alignment was suboptimal."
+        )
+
+    # apply the rigid transformation on the chains described in `apply_rigid_transformation_to_dynamic_chain_ids` argument
+    transformed_dynamic_atom_pos = {}
+    for chain_id, prot in apply_rigid_on_dynamic_chains.items():
+        _atom_pos_orig_shape = prot["atom14_gt_positions"].shape
+        _atom_pos_flat = prot["atom14_gt_positions"].reshape(-1, 3)
+        _atom_pos_flat_transformed = np.dot(_atom_pos_flat, rot_matrix) + trans_matrix
+        _atom_pos_transformed = _atom_pos_flat_transformed.reshape(
+            *_atom_pos_orig_shape
+        )
+        transformed_dynamic_atom_pos[chain_id] = _atom_pos_transformed
+
+        # transformed_dynamic_atom_pos[chain_id] = prot['atom14_gt_positions']
+
+    save_structure_file(
+        output_filename_extensionless=output_pdb_filename_extentionless,
+        pdb_id="unknown",
+        chain_to_atom14=transformed_dynamic_atom_pos,
+        chain_to_aa_str_seq={
+            chain_id: apply_rigid_on_dynamic_chains[chain_id]["aasequence_str"]
+            for chain_id in apply_rigid_on_dynamic_chains.keys()
+        },
+        chain_to_aa_index_seq={
+            chain_id: apply_rigid_on_dynamic_chains[chain_id]["aatype"]
+            for chain_id in apply_rigid_on_dynamic_chains.keys()
+        },
+        save_cif=False,
+        mask=None,  # TODO: check
+    )
+
+    # apply_on_atom_pos = apply_rigid_on_dynamic_concat['atom_positions']
```

Comment:
Do you need the commented out code below?

Reply:
deleted